src/hardware/adlib_gold.cpp

```diff
@@ -0,0 +1,434 @@
+	for (auto &f : highshelf)
+		f.setup(sample_rate, cutoff_freq, gain_db, slope);
+}
+
+StereoProcessor::~StereoProcessor() {}
+
+constexpr auto volume_0db_value       = 60;
+constexpr auto shelf_filter_0db_value = 6;
+
+void StereoProcessor::Reset() noexcept
+{
+	ControlWrite(StereoProcessorControlReg::VolumeLeft, volume_0db_value);
+	ControlWrite(StereoProcessorControlReg::VolumeRight, volume_0db_value);
+	ControlWrite(StereoProcessorControlReg::Bass, shelf_filter_0db_value);
+	ControlWrite(StereoProcessorControlReg::Treble, shelf_filter_0db_value);
+
+	StereoProcessorSwitchFunctions sf = {};
+	sf.source_selector                = static_cast<uint8_t>(
+                StereoProcessorSourceSelector::Stereo1);
+	sf.stereo_mode = static_cast<uint8_t>(StereoProcessorStereoMode::LinearStereo);
+
+	ControlWrite(StereoProcessorControlReg::SwitchFunctions, sf.data);
+}
+
+void StereoProcessor::ControlWrite(const StereoProcessorControlReg reg,
+                                   const uint8_t data) noexcept
+{
+	auto calc_volume_gain = [](const int value) {
+		constexpr auto min_gain_db = -128.0;
+		constexpr auto max_gain_db = 6.0;
+		constexpr auto step_db     = 2.0;
+
+		auto val     = value - volume_0db_value;
+		auto gain_db = clamp(val * step_db, min_gain_db, max_gain_db);
+		return static_cast<float>(decibel_to_gain(gain_db));
+	};
+
+	auto calc_filter_gain_db = [](const int value) {
+		constexpr auto min_gain_db = -12.0;
+		constexpr auto max_gain_db = 15.0;
+		constexpr auto step_db     = 3.0;
+
+		auto val = value - shelf_filter_0db_value;
+		return clamp(val * step_db, min_gain_db, max_gain_db);
+	};
+
+	constexpr auto volume_control_width = 6;
+	constexpr auto volume_control_mask  = (1 << volume_control_width) - 1;
+
+	constexpr auto filter_control_width = 4;
+	constexpr auto filter_control_mask  = (1 << filter_control_width) - 1;
+
+	switch (reg) {
+	case StereoProcessorControlReg::VolumeLeft: {
+		const auto value = data & volume_control_mask;
+		gain.left        = calc_volume_gain(value);
+		DEBUG_LOG_MSG("ADLIBGOLD: Stereo: Final left volume set to %.2fdB (value %d)",
+		              gain.left,
+		              value);
+	} break;
+
+	case StereoProcessorControlReg::VolumeRight: {
+		const auto value = data & volume_control_mask;
+		gain.right       = calc_volume_gain(value);
+		DEBUG_LOG_MSG("ADLIBGOLD: Stereo: Final right volume set to %.2fdB (value %d)",
+		              gain.right,
+		              value);
+	} break;
+
+	case StereoProcessorControlReg::Bass: {
+		const auto value   = data & filter_control_mask;
+		const auto gain_db = calc_filter_gain_db(value);
+		SetLowShelfGain(gain_db);
+
+		DEBUG_LOG_MSG("ADLIBGOLD: Stereo: Bass gain set to %.2fdB (value %d)",
+		              gain_db,
+		              value);
+	} break;
+
+	case StereoProcessorControlReg::Treble: {
+		const auto value   = data & filter_control_mask;
+		const auto gain_db = calc_filter_gain_db(value);
+		SetHighShelfGain(calc_filter_gain_db(data));
+
```

Comment:
`gain_db` is never used; was the intent 
```cpp
SetHighShelfGain(calc_filter_gain_db(gain_db));
```
?

Reply:
great catch, @kklobe !

Reply:
Yeah, fixing in the latest PR 👍🏻